File: palindrome/palindrome.py
def is_palindrome_simple(s):
    return s == s[::-1]

def is_palindrome_two_pointers(s):
    left, right = 0, len(s) - 1
    while left < right:
        if s[left] != s[right]:
            return False
        left += 1
        right -= 1
    return True

def is_palindrome_clean(s):
    cleaned = ''.join(c.lower() for c in s if c.isalnum())
    return cleaned == cleaned[::-1]

print(is_palindrome_simple("racecar"))
print(is_palindrome_two_pointers("A man a plan a canal Panama"))
print(is_palindrome_clean("race a car"))

File: palindrome/palindrome.rs
fn is_palindrome(s: &str) -> bool {
    let chars: Vec<char> = s.chars().collect();
    let mut left = 0;
    let mut right = chars.len() - 1;
    
    while left < right {
        if chars[left] != chars[right] {
            return false;
        }
        left += 1;
        right -= 1;
    }
    true
}

fn is_palindrome_clean(s: &str) -> bool {
    let cleaned: String = s.chars()
        .filter(|c| c.is_alphanumeric())
        .map(|c| c.to_lowercase().next().unwrap())
        .collect();
    
    cleaned == cleaned.chars().rev().collect::<String>()
}

fn main() {
    println!("{}", is_palindrome("racecar"));
    println!("{}", is_palindrome_clean("A man, a plan, a canal: Panama"));
}